js/trackers.js

```diff
@@ -233,8 +231,48 @@ require.scopes.trackers = (function() {
         return match
     }
 
+    // add block/noblock decisions (per domain) to the trackers cache
+    // remove older entries if the cache is full
+    function addToCache (reqUrl, currLocation, trackerData) {
+        if (!settings.getSetting('trackerBlockingEnabled')) return
+        if (isFirstPartyRequest(currLocation, reqUrl)) return
+        if (cache.size > 3000) {
+            cache.delete(cache.keys().next().value)
+        }
+        generateCacheKey(reqUrl, currLocation).then((hashedKey) => {
+            if (trackerData) {
+                cache.set(hashedKey, trackerData)
+            }
+        })
+    }
+
+    function isCached (reqUrl, currLocation) {
+        return new Promise((resolve, reject) => {
+            if (!settings.getSetting('trackerBlockingEnabled') ||
+                isFirstPartyRequest(currLocation, reqUrl)) {
+                resolve({block: false})
+            }
+            generateCacheKey(reqUrl, currLocation).then((hashedKey) => {
+                const cachedData = cache.get(hashedKey)
+                resolve(cachedData)
+            })
+        })
+    }
+
+    function generateCacheKey (reqUrl, currLocation) {
+        reqUrl = encodeURIComponent(reqUrl)
+        currLocation = encodeURIComponent(currLocation)
+        // cached trackers lookups are scoped to a domain
+        const key = `domain=${utils.extractHostFromURL(currLocation)},reqUrl:${reqUrl}`
+        return new Promise((resolve, reject) => {
+            utils.hashSHA256(key).then((hashedKey) => resolve(hashedKey))
```

Comment:
1) why are we double hashing again? security of the domain? are we sure that's really an issue?
2) this is looking like you want to subclass Map with your special behavior of 3000 and rehashed keys.

Reply:
1. The crypto (aka double) -hash is for privacy. Otherwise this is a record of user history in cleartext. 
2. Sure, I would consider that approach.

An alternate approach that would incorporate both your points above and improve the implementation which I was considering last week and will put on table here -- make the cache its own class and abstract them away from this `trackers` module (which is publicly available as a global -- meaning a user's history could theoretically be discovered by hashing the domain and tracker url manually and querying `trackers.isCached(key)` from the console. A highly unlikely scenario but still plausible. Either way I like the idea of subclassing Map/cache. I would like to leave the hashSHA256() func as a util so it can be repurposed in other places.

Reply:
(... meaning that util.hashSHA256() would be set to a convenience method of the Map subclass but also still available as util.hashSHA256() for other pieces of code that might want to utilize it)

Reply:
1. I'm just asking if we're really sure that's the case, a real risk. we don't want to store such a thing offline anyway and I think you'd want to prevent that from happening if it was, for performance reasons alone. Not saying don't do it, I'm asking how we know it's a real concern.

Reply:
I like your suggest re: caching per page so that invalidates the discussion about history. Thankfully!